src/rust/src/x509/verify.rs

```diff
@@ -180,6 +215,92 @@ self_cell::self_cell!(
     }
 );
 
+#[pyo3::pyclass(
+    frozen,
+    name = "ClientVerifier",
+    module = "cryptography.hazmat.bindings._rust.x509"
+)]
+struct PyClientVerifier {
+    policy: OwnedPolicy,
+    #[pyo3(get)]
+    store: pyo3::Py<PyStore>,
+}
+
+impl PyClientVerifier {
+    fn as_policy(&self) -> &Policy<'_, PyCryptoOps> {
+        &self.policy.borrow_dependent().0
+    }
+}
+
+#[pyo3::pymethods]
+impl PyClientVerifier {
+    #[getter]
+    fn validation_time<'p>(&self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        datetime_to_py(py, &self.as_policy().validation_time)
+    }
+
+    #[getter]
+    fn max_chain_depth(&self) -> u8 {
+        self.as_policy().max_chain_depth
+    }
+
+    fn verify(
+        &self,
+        py: pyo3::Python<'_>,
+        leaf: pyo3::Py<PyCertificate>,
+        intermediates: Vec<pyo3::Py<PyCertificate>>,
+    ) -> CryptographyResult<pyo3::Py<pyo3::types::PyTuple>> {
+        let policy = self.as_policy();
+        let store = self.store.get();
+
+        let chain = cryptography_x509_verification::verify(
+            &VerificationCertificate::new(
+                leaf.get().raw.borrow_dependent().clone(),
+                leaf.clone_ref(py),
+            ),
+            intermediates.iter().map(|i| {
+                VerificationCertificate::new(
+                    i.get().raw.borrow_dependent().clone(),
+                    i.clone_ref(py),
+                )
+            }),
+            policy,
+            store.raw.borrow_dependent(),
+        )
+        .map_err(|e| VerificationError::new_err(format!("validation failed: {e:?}")))?;
+
+        let py_chain = pyo3::types::PyList::empty(py);
+        for c in &chain {
+            py_chain.append(c.extra())?;
+        }
+
+        // NOTE: These `unwrap()` cannot fail, since the underlying policy
+        // enforces the presence of a SAN and the well-formedness of the
+        // extension set.
+        let leaf_san = &chain[0]
+            .certificate()
+            .extensions()
+            .unwrap()
+            .get_extension(&SUBJECT_ALTERNATIVE_NAME_OID)
+            .unwrap();
```

Comment:
Flagging: these unwraps are pretty gross, but we need them because we don't have an easy way to propagate the fact that the SAN/extensions were already enforced up here.